### PR TITLE
headscale: make module conform with rfc42 and upstream (0.17.1) config.

### PR DESCRIPTION
**Pull request by kradalby**
###### Description of changes

This PR changes the `headscale` module to conform with RFC42 as it had already drifted from upstream configuration and updates headscale to version 0.17.1.

There are currently multiple commits, and might be some more after comments, I'll squash it up before merge.

This PR contains the changes from #203694.
This PR supersede and closes #167513.

###### Things done

1. Formatted the code with `nixpkgs-fmt`
2. Moved all settings to `settings` key, mirroring the naming from upstream
3. Marked all renamed module settings

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Signed-off-by: Kristoffer Dalby <kristoffer@dalby.cc>

### CODE REVIEW

**Comment by kradalby** [comment]
Doc build failure seem to be related to the `redmine` module and not `headscale.

File: nixos/modules/services/networking/headscale.nix
Comment (06kellyjac) [inline nixos/modules/services/networking/headscale.nix:478]: spaces should remain. Same in other places

`nixpkgs-fmt`

File: nixos/doc/manual/release-notes/rl-2305.section.md
Comment (06kellyjac) [inline nixos/doc/manual/release-notes/rl-2305.section.md:43]: `[]` probably needs to be filled & the link updated to not be `grafana`

Comment (06kellyjac) [reply nixos/doc/manual/release-notes/rl-2305.section.md:43]: This list is pretty short so it's bound to happen right now but the advice is to put the item in at a random point in this list to avoid merge conflicts

**Review by 06kellyjac — COMMENTED** [review]
generally looks good. Added some comments

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix]: Addressed.

File: nixos/doc/manual/release-notes/rl-2305.section.md
Comment (kradalby) [reply nixos/doc/manual/release-notes/rl-2305.section.md]: Addressed

File: nixos/modules/services/networking/headscale.nix
Comment (Misterio77) [inline nixos/modules/services/networking/headscale.nix:423]:
```suggestion
        ${optionalString (cfg.settings.db_password_file != null) ''
          export HEADSCALE_DB_PASS="$(head -n1 ${escapeShellArg cfg.settings.db_password_file})"
```

File: nixos/modules/services/networking/headscale.nix
Comment (Misterio77) [inline nixos/modules/services/networking/headscale.nix:192]:
```suggestion
            db_password_file = mkOption {
```

**Review by Misterio77 — CHANGES_REQUESTED** [review]
Overall, looks great!

Just pulled it and tested on my server, here's a few small errors I found:

**Comment by RaitoBezarius** [comment]
Blocked on #203952.

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix]: fixed

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix]: fixed

**Comment by kradalby** [comment]
@Misterio77 How do you feel about me cherry picking your two relevant commits (f69b63ebb860312386569732124062a1c2ebc8be and f76bebb005a87b8e3d259f936195660c4f36cdff) onto this one?

**Comment by Misterio77** [comment]
Feel free! :)

**Comment by Misterio77** [comment]
Also, I'd love to help maintain the package and/or module. Feel free to add me as a maintainer if you folks could use some help.

**Review by Misterio77 — CHANGES_REQUESTED** [review]

**Comment by kradalby** [comment]
I picked @Misterio77's commits from #203694 onto this PR so we have a consistent change with since the new module and the new version depend on each other. 

I also added @Misterio77 as a headscale maintainer.

I believe this is good for review now, and I can squash it all up when someone has approved it.

**Review by Misterio77 — APPROVED** [review]
LGTM!

**Comment by kradalby** [comment]
Squashed up and ready to merge as long as people are happy, not sure if we need more approvers?

**Review by Misterio77 — APPROVED** [review]

**Comment by ghuntley** [comment]
@kradalby 👋 just stumbled upon this via a rando convo with Juan (I should have searched before hand tbh) 

Over at https://github.com/NixOS/nixpkgs/pull/204191 is my bump which I'll close out.

From what I can see rn there's three things missing from this pull-request rn

- [ ]  Implement breaking change `noise.private_key_path has been added and is required for the new noise protocol.`
- [ ]  Implement breaking change https://github.com/juanfont/headscale/pull/768
- [ ]  Implement https://github.com/juanfont/headscale/pull/905

Also consider applying backport label due to security fix https://github.com/juanfont/headscale/pull/823

File: nixos/tests/headscale.nix
Comment (ghuntley) [inline nixos/tests/headscale.nix:10]: Thank you for adding these tests.

File: pkgs/servers/headscale/default.nix
Comment (ghuntley) [inline pkgs/servers/headscale/default.nix:48]:
```suggestion
    maintainers = with maintainers; [ nkje jk kradalby misterio77 ghuntley ];
```

File: nixos/modules/services/networking/headscale.nix
Comment (ghuntley) [inline nixos/modules/services/networking/headscale.nix:216]:
```suggestion

            log.format = mkOption {
              type = types.str;
              default = "text";
              description = lib.mdDoc ''
                headscale log format.
              '';
              example = "json";
            };

```

File: nixos/modules/services/networking/headscale.nix
Comment (ghuntley) [inline nixos/modules/services/networking/headscale.nix:225]:
```suggestion


              override_local_dns = mkOption {
                type = types.bool;
                default = false;
                description = lib.mdDoc ''
                  Whether to use [Override local DNS](https://tailscale.com/kb/1054/dns/).
                '';
                example = true;
              };

```

**Comment by Misterio77** [comment]
Hey!

I was wondering if we should include all options like that. We're using freeform options following RFC42 but, [according to it](https://github.com/NixOS/rfcs/blob/45b76f20add8d7e0e1d0bc0eed9357b067c899a6/rfcs/0042-config-option.md#valuable-options), we should probably only explicitly include options that are either important, required for the program to run, or are secrets (so that the user can use a file instead).

**Comment by ghuntley** [comment]
>  explicitly include options that are either important

- override_local_dns is a key feature so my vote is to include?
- we have `log.level` ... undecided if to include `log.format`

CleanShot 2022-12-03 at 17 08 29@2x

A downside of RFC42 is it reduces discovery of options. I often head to https://search.nixos.org -> options and search for config options. Having items listed there highlights that the option is configurable.

**Comment by kradalby** [comment]
@ghuntley Thanks!, I will get in the changes a bit later today.

On backport: I am happy to have it backported, "open question", this will break people who has already upgraded and use headscale, what is NixOS's policy for that?

On adding all options: I am not against adding all options per sei, as I appreciate `search.nixos.org` a lot (Really I would love Nix to be _the way_ to configure headscale and the defacto standard). I like the way its documented and makes it discoverable. But I am worried about drift if I forget, or an option goes unnoticed by other maintainers. 

In any case, I do not have a strong opinion, and I'm happy to _try_ to keep it up to date, and with the new format, we allow passthrough of options so it should at least not block updates.

@Misterio77 @ghuntley @06kellyjac

File: pkgs/servers/headscale/default.nix
Comment (kradalby) [inline pkgs/servers/headscale/default.nix:19]:
```suggestion
  checkFlags = [ "-short" ];
  
  tags = ["ts2019"];
```

I forgot this, we need to add this tag to enable the "old" protocol.

**Review by Misterio77 — APPROVED** [review]

**Comment by viq** [comment]
Does anything else need to happen for this to be merged?

**Comment by viq** [comment]
@kradalby @Misterio77 @ghuntley @NKJe @06kellyjac Does anything else need to happen for this to be merged?

**Comment by kradalby** [comment]
@viq should not be, or I seem to have to resolve some new conflicts, but other than that its been ready a few weeks.

**Comment by K900** [comment]
I'm interested in getting this in, so ping me when you resolve conflicts and I can merge.

**Comment by kradalby** [comment]
@K900 @viq Resolved.

File: nixos/modules/services/networking/headscale.nix
Comment (K900) [inline nixos/modules/services/networking/headscale.nix:235]: Any particular reason we're setting this to false when upstream defaults to true? This seems to break MagicDNS by default (which may or may not be an upstream bug).

**Pull request merged**
